### PR TITLE
add timestamp support

### PR DESCRIPTION
**Pull request by kevinhu-nv**
# What does this PR do ?

Add support for parsing timestamp data for training.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Filter out transcription with text by regex
- Tokenize word by word and insert start and end timestamps for each word

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/tokenizers/canary_tokenizer.py [hunk 1]
```diff
@@ -73,7 +73,53 @@ def text_to_ids(self, text, lang_id) -> list[int]:
...
         if text.endswith(CANARY_EOS):
-            return super().text_to_ids(text[: -len(CANARY_EOS)], lang_id) + [self.eos_id]
+            text_no_eos = text[: -len(CANARY_EOS)]
+            time_pattern = re.compile(r"<\|\d+\|>")
+            time_text = "".join(time_pattern.findall(text_no_eos))
+            has_no_timestamp = not time_text
+            if has_no_timestamp:
+                return super().text_to_ids(text_no_eos, lang_id) + [self.eos_id]
+            else:
+                word_text = time_pattern.sub("", text_no_eos).strip()
```
Comment (pzelasko) [inline nemo/collections/common/tokenizers/canary_tokenizer.py:83]: please refactor into a method `def _text_with_timestamps_to_ids(...)` and use `return self._text_with_timestamps_to_ids(...)` here

File: nemo/collections/common/tokenizers/canary_tokenizer.py [hunk 2]
```diff
@@ -73,7 +73,53 @@ def text_to_ids(self, text, lang_id) -> list[int]:
...
+                        time_index += 2
+                # Ensure the last time_id is appended at the end
+                if time_index < len(time_ids):
+                    result_ids.append(time_ids[-1])
+                # Make sure the last time_id is appended only once
+                if time_index < len(time_ids) and result_ids[-1] != (time_ids[-1]):
+                    result_ids.append(time_ids[-1])
+                return result_ids + [self.eos_id]
+
         return super().text_to_ids(text, lang_id)
```
Comment (pzelasko) [inline nemo/collections/common/tokenizers/canary_tokenizer.py:123]: you missed this branch (when lang_id isn't CANARY_SPECIAL_TOKENIZER and the text doesn't end in EOS) - it also needs to support tokenizing timestamps

**Review by pzelasko — COMMENTED** [review]
Please add a unit test showing how a transcript with timestamp gets correctly tokenized (see tests/collections/asr/test_asr_multitask_model_bpe.py for reference tests you can use as a starting point)

File: nemo/collections/common/tokenizers/canary_tokenizer.py (hunk 1 above)
Comment (kevinhu-nv) [reply nemo/collections/common/tokenizers/canary_tokenizer.py]: Done.

File: nemo/collections/common/tokenizers/canary_tokenizer.py (hunk 2 above)
Comment (kevinhu-nv) [reply nemo/collections/common/tokenizers/canary_tokenizer.py]: Thanks for catching that - fixed.

**Comment by kevinhu-nv** [comment]
PTAL @pzelasko

**Review by pzelasko — APPROVED** [review]

**Review by pzelasko — APPROVED** [review]

**Comment by kevinhu-nv** [comment]
@titu1994 @stevehuang52 Could you take a look to trigger the workflow checks?

**Pull request merged**